### DISCUSSION TITLE
Option to remove “Safely Remove Hardware” System tray Icon

### DISCUSSION THREAD

**Discussion by chip33** [valinet/ExplorerPatcher#431] [Ideas]
In a quest to reduce the number of unused or unnecessary icons in the system tray I recently found [this article](https://www.raymond.cc/blog/hide-or-remove-safely-remove-hardware-tray-icon/).

Changing a registry value and running the systray.exe file to refresh the setting can completely remove (not hide) the Safely Remove Hardware icon.

The registry value will revert back to default when you restart. So you can reset the registry value and run systray when Windows is booted up.

You can use this batch file run at startup:

`reg add "HKCU\Software\Microsoft\Windows\CurrentVersion\Applets\SysTray" /v "Services" /t reg_dword /d 29 /f
systray`

edit:  Some users in the comments section of the linked article reported needing to add a small delay to the batch file.

Or in my case I used Resource Hacker to swap the XP style graphics with Windows 10 style in the freeware Icon Remover tool linked in the above article.

![hide-safely-remove-hardware](https://user-images.githubusercontent.com/48499740/142722098-09731f65-2527-43f9-9222-1b75e9b8f1db.png)

You can still just right click on any USB device in My Computer and select “Eject”.

Being Taskbar related and an improving tweak I have found to work well with ExplorerPatcher I mention it here as a possible addition.
Anything to unclutter the system tray is good for me.

**Comment by chip33** [comment]
Hi, 
I was reading about how new additions to the program can cause work for valinet down the line due to issues raised correctly or incorrectly etc. and how this could lead to a lessening of new functions being added , especially as this forum is very active indeed.
The tweak I have described above seems one that is very useful to myself albeit a modest one. 
Apart from the ancient Windows XP era app linked above I have not found this duplicated functionality in any other program.
Also, naively perhaps I cannot see how enabling this is likely to cause support issues.
It is just a fix  (enabled by a startup regedit) for MS not including this icon in the list of System Icons that can be turned off.
I have the XP app working well on my system so will be fine without it going in but it is also something that could be (almost) unique to EP.
Seems like there are strong valid reasons to reduce support time by not adding features...
So if we see an idea or feature suggested in the forum and we like it shouldn't we support that feature to show it was not added just as a niche case for a very few users.
Maybe if anyone else feels this would be a handy addition they could upvote or comment on this thread.
Thanks .

**Reply by valinet** [reply]
Hey, it's not a vote, we're not electing anything. I don't care about how badly requested a feature is, or not, I care about someone maintaining it and contributing it with complete and well tested code. Up votes don't help code it better and take out bugs etc. I am also not selling a product - I don't care about its popularity or to service users a certain way. This ships mostly as an implementation of my vision and of the things I use. As you see, I still haven't added the Windows 11 orb or rounded corners or accent everywhere as i do not use that. I know some people want them, but I personally may do them when I have time. Other than that, I am open to well defined contributions for them, but knowing I don't use them, I am not interested into entering into another "support war". Also, as an example, idk how used the Windows 10 switcher is and frankly I do not care, I include it because it fits my vision and needs and I sometimes dedicate time on fixing and improving it. If people agree with that I ship, fine, they can use, if they feel like something else should be added and have a solid plan for how that should be supported, fine as well, if they don't but still feel like including something, forking is always an option. There are plenty of options to choose for this, let's think what benefits all parties involved, not a certain group. That's the blessing and cursing of projects as they grow larger and larger. It's hard to maintain the dynamism of the early versions when there still only one developer involved that has to spend a lot of time polishing what's already in there, let alone add new stuff. I don't want the fiasco in the other posts to perpetuate. It is not about adding features, being unique etc necessarily. This project is aimed at offering en educational insight and delivering a productive working environment, similar to Windows 10.

Thanks

**Reply by chip33** [reply]
Hi, I appreciate the long reply.
What do you think of it in terms of the regedit is supplied and the general usefulness of having one less item in the system tray.
I can totally get if you think it just isn't worth the extra clutter in the Properties GUI or just that it doesn't appeal to you. Also I guess although it seems simple maybe there could be unforeseen negative side effects.
Anyway , thanks for replying.

**Reply by valinet** [reply]
I think the reply above made my point more than clear.

**Reply by chip33** [reply]
I wasn't quite sure but I am now :)

**Comment by chip33** [comment]
Hi, I have been trying to implement this in my own fork by studying your code.
All I have to do is change
 ```
[HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Applets\SysTray]
"Services"=dword:0000001f
```
to
`"Services"=dword:0000001d `
and restore it when unticked in Settings
and to run systray.exe to implement the new reg value.

I can add the option to settings:

```
[HKEY_CURRENT_USER\Software\ExplorerPatcher]
;b Hide Safely Remove Hardware button
"HideSafelyRemoveHardwarebutton"=dword:00000000
```

I can see how to then reference this in dllmain.c and I can run systray.exe but cannot quite work out how to toggle the non boolean dword data for "Services" and enforce it (if modded) at PC restart.

I have been looking to modify code that does a similar changing of registry data in dllmain.c but cannot get it to work, is it possible to point me towards the best code in there to adapt or give me a code snippet if you don't actually do this in the existing code?
